products/llm_observability/frontend/LLMObservabilityTraceScene.tsx

```diff
@@ -189,72 +190,151 @@ function NestingGroup({ level = 0, children }: { level?: number; children: React
+    return (
+        <NestingGroup level={1}>
+            {tree.map(({ event, children }) => (
+                <React.Fragment key={event.id}>
+                    <TraceNode
+                        topLevelTrace={trace}
+                        item={event}
+                        isSelected={!!selectedEventId && selectedEventId === event.id}
+                    />
+                    <li>
+                        {children && (
+                            <RecursiveTreeDisplay tree={children} trace={trace} selectedEventId={selectedEventId} />
+                        )}
+                    </li>
+                </React.Fragment>
+            ))}
+        </NestingGroup>
+    )
+}
 
+function EventContentDisplay({
+    input,
+    output,
+    raisedError,
+}: {
+    input: unknown
+    output: unknown
+    raisedError?: boolean
+}): JSX.Element {
     return (
-        <li key={item.id} className="mt-0.5">
-            <Link
-                to={urls.llmObservabilityTrace(topLevelTrace.id, {
-                    event: item.id,
-                    timestamp: removeMilliseconds(topLevelTrace.createdAt),
-                })}
-                className={classNames(
-                    'flex flex-col gap-1 p-1 text-xs rounded hover:bg-accent-primary-highlight',
-                    isSelected && 'bg-accent-primary-highlight'
-                )}
-            >
-                <div className="flex flex-row flex-wrap items-center gap-1.5">
-                    <EventTypeTag event={item} size="small" />
-                    <span>
-                        {'properties' in item
-                            ? `${item.properties.$ai_model} (${item.properties.$ai_provider})`
-                            : item.traceName}
-                    </span>
+        <LLMInputOutput
+            inputDisplay={
+                <div className="p-2 text-xs border rounded bg-[var(--bg-fill-secondary)]">
+                    {isObject(input) ? (
+                        <JSONViewer src={input} collapsed={4} />
+                    ) : (
+                        <span className="font-mono">{JSON.stringify(input ?? null)}</span>
+                    )}
                 </div>
-                <div className="flex flex-row flex-wrap text-muted items-center gap-1.5">
-                    <LemonTag type="muted">{formatLLMLatency(latency)}</LemonTag>
-                    {(usage != null || totalCost != null) && (
-                        <span>
-                            {usage}
-                            {usage != null && totalCost != null && <span>{' / '}</span>}
-                            {totalCost != null && formatLLMCost(totalCost)}
-                        </span>
+            }
+            outputDisplay={
+                <div
+                    className={cn(
+                        'p-2 text-xs border rounded',
+                        !raisedError ? 'bg-[var(--bg-fill-success-tertiary)]' : 'bg-[var(--bg-fill-warning-tertiary)]'
```

Comment:
Let's use red to signal an error:
```suggestion
                        'p-2 text-xs border rounded',
                        !raisedError ? 'bg-[var(--bg-fill-success-tertiary)]' : 'bg-[var(--bg-fill-danger-tertiary)]'
```

Reply:
I used the danger color initially, but the contrast ratio is not great and is hard to read. When you look at those spans for a little longer, you actually want it to be plain and readable. I've thought about changing the header color instead of applying a background, as it serves the same indicative purpose, but the content is significantly easier to read.

Reply:
Error

<img width="907" alt="Screenshot 2025-01-28 at 18 20 44" src="https://github.com/user-attachments/assets/91d6fe00-f6f7-4c4b-9924-755b1f3fdaf1" />
<img width="907" alt="Screenshot 2025-01-28 at 18 21 00" src="https://github.com/user-attachments/assets/d640116a-6a39-4a32-99c9-89296f7ad31a" />

Danger

<img width="912" alt="Screenshot 2025-01-28 at 18 21 44" src="https://github.com/user-attachments/assets/7cb777e8-fbe6-45f8-95bf-caf2855fd768" />
<img width="898" alt="Screenshot 2025-01-28 at 18 21 53" src="https://github.com/user-attachments/assets/b0f71de2-f781-4c6c-9b1a-bd06aeeff4a3" />

I'll switch to red, as I think it looks better now. Let's see.